usplash.8
.TH usplash "8" "June 2006" Ubuntu
.\"
.SH NAME
usplash \- userspace bootsplash daemon
.\"
.SH SYNOPSIS
\fBusplash\fR [\fB-c\fR] [\fB-v\fR] [\fB-x\fR \fInum\fR] [\fB-y\fR \fInum\fR]
.\"
.SH DESCRIPTION
Uses the framebuffer to provide a graphical bootsplash screen from
userspace during system boot and shutdown.
As well as a static image, the screen includes a progress bar and text
updated by commands received from
.BR usplash_write (8).
.\"
.SH OPTIONS
.TP
.B -c
Causes \fBusplash\fR to switch to a new console (vt8) before displaying
the splash screen, rather than writing over the current one.
.TP
.B -v
Put usplash into verbose mode where more messages will be shown.
Without this flag only urgent messages will be displayed.
.TP
.BI -x " num"
Set the x-axis resolution to use for the splash screen to
.IR num .
.TP
.BI -y " num"
Set the y-axis resolution to use for the splash screen to
.IR num .
.\"
.SH COMMANDS
See
.BR usplash_write (8)
for a list of the commands that can be sent.
.\"
.SH AUTHOR
Written by Matthew Garrett <mjg59@srcf.ucam.org>
.\"
.SH REPORTING BUGS
Report bugs at
.br
http://launchpad.net/distros/ubuntu/+source/usplash/+filebug
.\"
.SH COPYRIGHT
Copyright \(co 2006 Canonical Ltd.
.br
Copyright \(co 2006 Dennis Kaarsemaker <dennis@kaarsemaker.net>
.br
Copyright \(co 2005 Matthew Garrett <mjg59@srcf.ucam.org>

This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
.\"
.SH SEE ALSO
.BR usplash_write (8)


debian/changelog
usplash (0.5.2-1) unstable; urgency=low

  * New upstream release.
  * debian/control:
    - Change libgd2-dev build-dependency to libgd2-noxpm-dev;
    - Change libusplash0 dependency to make it bin-NMU friendly;
    - Remove a duplicate entry of section field at libusplash-dev;
  * debian/lsb-base-logging-ubuntu.sh: Add a hack while we don't migrate
    to the proposed splash methods to avoid to patch KDM and GDM.

 -- Otavio Salvador <otavio@ossystems.com.br>  Sun, 17 Jun 2007 18:02:06 -0300

usplash (0.4-43-3) unstable; urgency=low

  * debian/control: Fix section of libusplash-dev.
    Mirror active Maintainers and Uploaders from usplash-theme-debian.
    Replaces and Provides bootsplash. (closes: #422761)
  * lsb-base-logging.sh: Fix log_action_end_msg return code to be consistent
    with lsb-base. Thanks martin f krafft <madduck@debian.org> for patch.
    (closes: #422624)
  * initramfs/scripts/init-top/usplash: Disable splash too on single bootarg.

 -- maximilian attems <maks@debian.org>  Tue, 08 May 2007 14:20:50 +0200

usplash (0.4-43-2) unstable; urgency=low

  * initramfs/scripts/init-top/usplash: Enable splash by default on
    installation - patch by David Härdeman. (closes: #397954, #383327)
  * debian/control: Add usplash-theme-debian | usplash-theme depends
    for usplash auto-installed theme.

 -- maximilian attems <maks@debian.org>  Mon, 07 May 2007 18:13:03 +0200

usplash (0.4-43-1) unstable; urgency=low

  [ maximilian attems ]
  Release overlong fun duty

  * new upstream release (svga i386, theming support, input, packaging)
    drop all merged patches plus quilt rules, also the 04-bogl-dpkg-arg.patch,
    as you neeed newer debhelper than sarge version. It no longer leaves
    /var/lib/usplash around.  (closes: 380750, 383249, 397970, 382675)
  * Add lintian override for libusplash.so.0
  * Be vendor neutral in an ubuntu called lsb script. (closes: 384812)
  * debian/usplash.init: Add dep info. (closes: 386349)
    Thanks Petter Reinholdtsen <pere@hungry.com>
  * debian/usplash.init: Only run on stop with splash bootarg.
  * debian/control: Remove dependency on policycoreutils.
  * usplash-theme.h: Add input patch from David.

  [ Otavio Salvador ]
  * Change package to non-native;
  * Revert a usplash-theme.h change that isn't supported on the current
    code;
  * Avoid console-tools and kbd starting when usplash is running;
  * Add myself on uploaders.

 -- Otavio Salvador <otavio@debian.org>  Thu, 15 Mar 2007 15:51:15 -0300

usplash (0.3e) unstable; urgency=low

  * Add 05-usplash_write-quiet.patch from Petter Reinholdtsen
    <pere@hungry.com>. Gets rid of error messages on splash=false boots.

 -- maximilian attems <maks@sternwelten.at>  Fri,  4 Aug 2006 10:04:59 +0200

usplash (0.3d) unstable; urgency=low

  * Add 03-bogl-no-PAGESIZE.patch to the patches/series file to really fix
    the powerpc ftbfs.
  * bogl/Makefile: Ease sarge backports, build with DEB_BUILD_ARCH in
    04-blog-dpkg-arg.patch. Thanks Petter Reinholdtsen <pere@hungry.com>.
    (closes: 380745)
  * debian/usplash.init: Workaround missing ubuntu gdm patches. Call usplash
    in init stop. Thanks Petter Reinholdtsen <pere@hungry.com>.
    (closes: 381009)

 -- maximilian attems <maks@sternwelten.at>  Wed,  2 Aug 2006 09:28:22 +0200

usplash (0.3c) unstable; urgency=low
  * debian/usplash.postinst: Readd rm_default_artwork for migration from ubuntu.
  * Add 03-bogl-no-PAGESIZE.patch. (closes: 380618)

 -- maximilian attems <maks@sternwelten.at>  Mon, 31 Jul 2006 23:56:26 +0200

usplash (0.3b) unstable; urgency=low

  * debian/copyright: Add notice about .bdf files.

 -- maximilian attems <maks@sternwelten.at>  Sun, 30 Jul 2006 21:50:42 +0200

usplash (0.3a) unstable; urgency=low

  [ Petter Reinholdtsen ]
  * Initial upload to Debian (Closes: #356193)
  * usplash_write.c: Fix typo in fifo write code, 00-ssize-typo.patch.
  * usplash.c: Improve error messages when usplash fail to start,
    01-bogl-error.patch.
  * usplash.h: Improve comment for USPLASH_FIFO, 02-header-comment.patch.
  * Make sure postrm call update-rc.d to remove the init.d script.

  [ maximilian attems ]
  * Add the Debian kernel team as Maintainer. Petter and myself as Uploaders.
  * Add versioned dependency to initramfs-tools 0.72, which adds the fb boot
    scripts.
  * Standards-Version 3.7.2.0 without needed changes.
  * debian/copyright: Add pointer to latest Ubuntu release, small rephrasing.
  * Use a Debian version string, as there is no separate archive.
  * debian/usplash.postinst: Check if update-initramfs is really installed.
    Remove old unused rm_default_artwork().
  * debian/usplash.postrm: Check if update-initramfs is really installed.
  * debian/patches: Split out Petter's changes as patches.
  * debian/rules: Add quilt handling for the patches.
  * debian/control: Add quilt build dependency.
  * .bzr, bogl/.cvsignore: Remove version control stuff in source package.
  * Add lsb-base-logging-ubuntu.sh from ubuntu lsb (3.1-10ubuntu2).
  * debian/rules: Install lsb-base-logging.sh.
  * usplash.lintian: Add lintian override for /etc/lsb-base-logging.sh,
    Debian lsb-base allows to set there the lsb functions.
  * debian/rules: Install lintian override.
  * debian/control: Provides bootsplash as discussed with otavio.

 -- maximilian attems <maks@sternwelten.at>  Sat, 29 Jul 2006 17:16:04 +0200


usplash.c
/* usplash
 *
 * Copyright © 2006 Canonical Ltd.
 * Copyright © 2006 Dennis Kaarsemaker <dennis@kaarsemaker.net>
 * Copyright © 2005 Matthew Garrett <mjg59@srcf.ucam.org>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301 USA
 */

#include <linux/vt.h>

#include <sys/select.h>
#include <sys/time.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/ioctl.h>

#include <dlfcn.h>
#include <errno.h>
#include <fcntl.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <assert.h>
#include <signal.h>
#include <errno.h>

#include "usplash_backend.h"
#include "usplash.h"
#include "libusplash.h"

static int main_loop(void);
static int read_command(void);
static int parse_command(const char *string, size_t len);
static void do_animate(int ignore);

/* Number of seconds to wait for a command before exiting */
static int timeout = 15;

/* Are we pulsating the throbber? */
static int pulsating = 0;
static int framedelay = 40000;
static int cycles = 0;
static int cycle_timeout = 375;

static int fifo_fd;


int main(int argc, char *argv[])
{
	int i, ret = 0;
	int xres = 0;
	int yres = 0;
	int verbose = 0;
	struct timeval t1 = {.tv_sec = 0,.tv_usec = 40000 };
	struct timeval t2 = {.tv_sec = 0,.tv_usec = 40000 };
	struct itimerval iv = {.it_interval = t1,.it_value = t2 };

	for (i = 1; i < argc; i++) {
		if (!strcmp(argv[i], "-c")) {
			int vt_fd = open("/dev/tty8", O_RDWR);
			switch_console(8, vt_fd);
			close(vt_fd);
		} else if (!strcmp(argv[i], "-v")) {
			verbose = 1;
		} else if (!strcmp(argv[i], "-x")) {
			xres = atoi(argv[++i]);
		} else if (!strcmp(argv[i], "-y")) {
			yres = atoi(argv[++i]);
		} else {
			exit(1);
		}
	}

	if (chdir(USPLASH_DIR) < 0) {
		perror("chdir");
		ret = 1;
		goto exit;
	}

	if (mkfifo(USPLASH_FIFO, S_IRUSR | S_IWUSR | S_IRGRP | S_IWGRP) <
	    0) {
		if (errno != EEXIST) {
			perror("mkfifo");
			ret = 1;
			goto exit;
		}
	}

	if (mkfifo(USPLASH_OUTFIFO, S_IRUSR | S_IWUSR | S_IRGRP | S_IWGRP)
	    < 0) {
		if (errno != EEXIST) {
			perror("mkfifo");
			ret = 1;
			goto exit;
		}
	}

	fifo_fd = open(USPLASH_FIFO, O_RDONLY | O_NONBLOCK);
	if (fifo_fd < 0) {
		perror("open");
		ret = 2;
		goto exit;
	}

	if (usplash_setup(xres, yres, verbose)) {
		fprintf(stderr, "screen init failed\n");
		ret = 3;
		goto exit;
	}

	signal(SIGALRM, do_animate);
	setitimer(ITIMER_REAL, &iv, NULL);

	clear_screen();
	clear_progressbar();
	clear_text();

	ret = main_loop();

	usplash_done();
      exit:
	/* While usplash is running, /dev/console is actually tty8
	 * so if we flip back to where we came from, we actually hide the
	 * very messages we're trying to show the user!
	 *
	 * The case of usplash terminating normally is already handled
	 * in the init script with an explicit chvt.
	 */
//      usplash_restore_console ();
	return ret;
}

static void do_animate(int ignore)
{
	animate_step(pulsating);
}

static int main_loop(void)
{
	struct timeval tv;

	for (;;) {
		fd_set rfds;
		int retval;

		FD_ZERO(&rfds);
		FD_SET(fifo_fd, &rfds);

		tv.tv_sec = timeout;
		tv.tv_usec = 0;

		retval = select(fifo_fd + 1, &rfds, NULL, NULL, &tv);

		if (retval < 0 && errno != EINTR) {
			/* Error */
			return 1;
		} else if (retval < 0 && errno == EINTR) {
			/* Count cycles for timeout */
			cycles++;
			if (cycles >= cycle_timeout)
				return 0;
		} else if (retval > 0) {
			/* Data available */
			int ret;
			cycles = 0;
			ret = read_command();
			if (ret)
				return ret;
		}
	}

	/* Not reached */
	return 0;
}

static int read_command(void)
{
	static char buf[4096], *ptr;
	static size_t buflen;
	static ssize_t len;

	len = read(fifo_fd, buf + buflen, sizeof(buf) - buflen);
	if (len < 0) {
		if (errno != EAGAIN) {
			/* Try opening again */
			close(fifo_fd);
			fifo_fd =
			    open(USPLASH_FIFO, O_RDONLY | O_NONBLOCK);
			if (fifo_fd < 0)
				return 2;
		}

		return 0;
	} else if (len == 0) {
		/* Reopen to see if there's anything more for us */
		close(fifo_fd);
		fifo_fd = open(USPLASH_FIFO, O_RDONLY | O_NONBLOCK);
		if (fifo_fd < 0)
			return 2;

		return 0;
	}

	buflen += len;
	while ((ptr = memchr(buf, '\0', buflen)) != NULL) {
		int ret;

		ret = parse_command(buf, ptr - buf);
		if (ret)
			return ret;

		/* Move the rest of the buffer up */
		buflen -= ptr - buf + 1;
		memcpy(buf, ptr + 1, buflen);
	}

	return 0;
}

static int parse_command(const char *string, size_t len)
{
	const char *command;
	size_t commandlen;

	command = string;
	commandlen = strncspn(string, len, " ");

	if (string[commandlen] == ' ') {
		string += commandlen + 1;
		len -= commandlen + 1;
	} else {
		len = 0;
	}


	if (!strncmp(command, "QUIT", commandlen)) {
		return 1;

	} else if (!strncmp(command, "TIMEOUT", commandlen)) {
		timeout = atoi(string);
		cycle_timeout = (1000000 / framedelay) * timeout;

	} else if (!strncmp(command, "CLEAR", commandlen)) {
		clear_text();

	} else if (!strncmp(command, "TEXT", commandlen)) {
		draw_text(string, len);

	} else if (!strncmp(command, "TEXT-URGENT", commandlen)) {
		draw_text_urgent(string, len);

	} else if (!strncmp(command, "STATUS", commandlen)) {
		draw_status(string, len, 0);

	} else if (!strncmp(command, "SUCCESS", commandlen)) {
		draw_status(string, len, 1);

	} else if (!strncmp(command, "FAILURE", commandlen)) {
		draw_status(string, len, -1);

	} else if (!strncmp(command, "PROGRESS", commandlen)) {
		pulsating = 0;
		draw_progressbar(atoi(string));

	} else if (!strncmp(command, "PULSATE", commandlen)) {
		pulsating = 1;

	} else if (!strncmp(command, "INPUT", commandlen)) {
		return handle_input(string, len, 0);

	} else if (!strncmp(command, "INPUTQUIET", commandlen)) {
		return handle_input(string, len, 1);

	} else if (!strncmp(command, "INPUTENTER", commandlen)) {
		return handle_input(string, len, 2);
	}

	return 0;
}
